RATIONALE.md

```diff
@@ -1,5 +1,117 @@
+
+A small amount of stable fields (<=20) can be optimally done in an ABI model,
+even manually. In stable/small case, there isn't a lot of glue code needed and
+there is little maintenance to those functions over time. It is the most
+performant and efficient way to communicate.
+
+However, this doesn't match the use case of Kubernetes. The node and pod models
+include several hundred data types, with nearly a thousand fields. To deal with
+a model this large would require a code generator. Even if such a code
+generator were sourced or made, it would require automatic conversion from the
+incoming proto model used on the host, as the conversion logic would be too
+large to maintain manually. Many other complications would follow suit,
+including an amplified number of host calls. In summary, a usable ABI binding
+would be an effort larger than the scheduler plugin itself.
+
+The path of least resistance is a marshalling approach. Almost all parameters
+of size used in scheduler plugins are generated protobuf model types. While not
+performant, it is possible to have the host marshal these objects and unmarshal
+them as needed on the guest. Further optimizations could be made as many of the
+underlying data do not change within a plugin lifecycle.
+
+For example, a function to get a pod could look like this on the guest:
+```webassembly
+(import "k8s.io/api" "pod" (func $get_pod
+  (param $buf i32) (param $buf_limit i32)
+  (result (; pod_len ;) i32)))
+```
+
+The host would marshal the entire pod argument to protobuf like so:
+```go
+func k8sApiPodFn(ctx context.Context, mod wazeroapi.Module, stack []uint64) {
+	buf := uint32(stack[0])
+	bufLimit := bufLimit(stack[1])
+
+	pod := filterArgsFromContext(ctx).pod
+	stack[0] = uint64(marshalIfUnderLimit(mod.Memory(), pod, buf, bufLimit))
+}
+```
+
+The main tradeoff to this approach is performance. The model is very large and
+the default unmarshaller creates a lot of garbage. We have some workarounds
+noted below, and also there are options for mitigation not yet implemented:
+
+* updating heavy objects only when they change
+    * this limits the occurrence count of decode performance
+* tracking of alternative encoding libraries which use protobuf IDL
+    * [polyglot][3] can generate code from protos and might automatically convert
+      protos to its more efficient representation.
+
+## How is `framework.CycleState` implemented in WebAssembly?
+
+`framework.CycleState` is a primarily a key value storage. Plugins store values
+in `PreFilter` or `PreScore` for use afterward. This is similar to Go context,
+except the keys must be strings and the values must implement `StateData`.
+
+While one instance of `framework.CycleState` is re-used for all plugins in a
+scheduling cycle, in practice, `StateData` are not shared. For this reason, we
+implement cycle state in the guest. The only responsibility of the wasm guest
+is to reset any state when `PreFilter` is called. All state is invisible from
```

Comment:
yes, and you reminded me something we discussed offline, but I forgot to add. Basically wasm memory can grow but never shrink.

So cleanup in wasm will not really "release resources", as the memory will never get smaller. However, it could possible help trigger a GC. I need to write this into the RATIONALE just as a side note about the impact of not having a "done" hook in the scheduler framework yet. Basically, I agree and want a reliable hook, though mainly what it buys us is an explicit clean hook vs implicit/heuristics (looking for an indirect signal like PreFilter being called)

Reply:
actually I will clarify in the notes that yes in the future someone could have a non-heap resource like a file handle, and that would be left open until the next call to PreFilter